src/app/rpmostree-ex-builtin-jigdo2commit.c

```diff
@@ -0,0 +1,366 @@
+    if (pkglist->len > 1)
+      {
+        g_print ("%u oirpm matches\n", pkglist->len);
+      }
+    g_ptr_array_set_size (pkglist, 1);
+    if (!rpmostree_context_set_packages (self->ctx, pkglist, cancellable, error))
+      return FALSE;
+    oirpm_pkg = g_object_ref (pkglist->pdata[0]);
+  }
+
+  g_print ("oirpm: %s (%s)\n", dnf_package_get_nevra (oirpm_pkg),
+           dnf_package_get_reponame (oirpm_pkg));
+
+  if (!rpmostree_context_download (self->ctx, cancellable, error))
+    return FALSE;
+
+  glnx_fd_close int oirpm_fd = -1;
+  if (!rpmostree_context_consume_package (self->ctx, oirpm_pkg, &oirpm_fd, error))
+    return FALSE;
+
+  g_autoptr(RpmOstreeJigdoAssembler) jigdo = rpmostree_jigdo_assembler_new_take_fd (&oirpm_fd, oirpm_pkg, error);
+  if (!jigdo)
+    return FALSE;
+  g_autofree char *checksum = NULL;
+  g_autoptr(GVariant) commit = NULL;
+  g_autoptr(GVariant) commit_meta = NULL;
+  g_autoptr(GVariant) pkgs = NULL;
+  if (!rpmostree_jigdo_assembler_read_meta (jigdo, &checksum, &commit, &commit_meta, &pkgs,
+                                  cancellable, error))
+    return FALSE;
+
+  g_print ("OSTree commit: %s\n", checksum);
+
+  { OstreeRepoCommitState commitstate;
+    gboolean has_commit;
+    if (!ostree_repo_has_object (self->repo, OSTREE_OBJECT_TYPE_COMMIT, checksum,
+                                 &has_commit, cancellable, error))
+      return FALSE;
+    if (has_commit)
+      {
+        if (!ostree_repo_load_commit (self->repo, checksum, NULL, &commitstate, error))
+          return FALSE;
+        if (!(commitstate & OSTREE_REPO_COMMIT_STATE_PARTIAL))
+          {
+            g_print ("Commit is already written, nothing to do\n");
+            return TRUE;  /* 🔚 Early return */
+          }
+      }
+  }
+
+  g_printerr ("TODO implement GPG verification\n");
+
+  g_auto(RpmOstreeRepoAutoTransaction) txn = { 0, };
+  if (!rpmostree_repo_auto_transaction_start (&txn, self->repo, FALSE, cancellable, error))
+    return FALSE;
+
+  if (!rpmostree_jigdo_assembler_write_new_objects (jigdo, self->repo, cancellable, error))
+    return FALSE;
+
+  if (!commit_and_print (self, &txn, cancellable, error))
+    return FALSE;
+
+  /* Download any packages we don't already have imported */
+  g_autoptr(GPtrArray) pkgs_required = g_ptr_array_new_with_free_func (g_object_unref);
+  const guint n_pkgs = g_variant_n_children (pkgs);
+  for (guint i = 0; i < n_pkgs; i++)
+    {
+      const char *name, *version, *release, *architecture;
+      const char *repodata_checksum;
+      guint64 epoch;
+      g_variant_get_child (pkgs, i, "(&st&s&s&s&s)",
+                           &name, &epoch, &version, &release, &architecture,
+                           &repodata_checksum);
+      // TODO: use repodata checksum, but probably only if covered by the ostree
+      // gpg sig?
+      DnfPackage *pkg = query_nevra (dnfctx, name, epoch, version, release, architecture, error);
+      if (!pkg)
+        return FALSE;
```

Comment:
We shouldn't require a package to be in the repos if we already have it imported, right?

Reply:
Yeah that's a very good point.  We want to support upgrading ancient systems.